File: python/101-150/150. Evaluate Reverse Polish Notation.py
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        stack = []
        for c in tokens:
            if c not in "+-*/":
                val = int(c)
            else:
                val2 = stack.pop()
                val1 = stack.pop()
                if c == "+":
                    val = val1 + val2
                if c == "-":
                    val = val1 - val2
                if c == "*":
                    val = val1 * val2
                if c == "/":
                    if val1 * val2 < 0 and val1 % val2 != 0:
                        val = val1 / val2 + 1
                    else:
                        val = val1 / val2
            stack.append(val)
        return stack[-1]
            

File: cpp/101-150/150. Evaluate Reverse Polish Notation.cc
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
    	unordered_map<string, function<int (int, int)>> map = {
    		{"+", [](int a, int b) {return a + b;}},
    		{"-", [](int a, int b) {return a - b;}},
    		{"*", [](int a, int b) {return a * b;}},
    		{"/", [](int a, int b) {return a / b;}}
    	};
        if (tokens.empty())
            return 0;
        stack<int> s;
        for (int i = 0; i < tokens.size(); i++) {
            if(map.find(tokens[i]) == map.end()){
            	s.push(stoi(tokens[i]));
            }
            else {
            	int op2 = s.top();
            	s.pop();
            	int op1 = s.top();
            	s.pop();
                s.push(map[tokens[i]](op1, op2));
            }
        }
        return s.top();
    }
};